### ISSUE TITLE
Version 1.0

### ISSUE THREAD

**Issue by andywer** [andywer/threads.js#100]
Hey everyone!
I finally came up with a prototype for v1.0 🚀

It's a complete rewrite of the library
* with a fully redesigned API
* focussing on promises and observables
* written completely in TypeScript

It's not completely implemented yet, but most of the basics are already there and do already work.

Find it here:
👉 **[Branch `v1`](https://github.com/andywer/threads.js/tree/v1)**

Please have a look at it and leave your feedback here in this issue 👇🙌

## Progress

- [x] Implement `spawn`
- [x] Implement `Worker`
- [x] Implement `Thread.*` utils
- [x] Implement `expose`
- [x] Implement platform: node 12 `worker_threads`
- [x] Implement platform: node < 12 via `tiny-worker`
- [x] Implement platform: web (browsers)
- [x] Implement transferable object support
- [x] Implement `Pool`
- [x] Add tests for usage with webpack
- [x] Implement cancelling pool tasks
- [x] Provide a convenient way to stream results
- [x] Finish documentation
- [x] Finish jsdocs
- [ ] Allow passing callbacks\* (#145)
- [ ] ~~Implement progress reporting~~\* (kind-of covered by observable support already)
- [ ] ~~Implement passing promises & observables from master to worker~~\* (rare use cases)

*\* Nice to have in v1.0. Might be postponed and implemented in v1.1 or so.*

**Comment by andywer** [comment]
@gtdminh @ItsEcholot @aaronbeall I know it's been a while, but we are finally getting close to a relaunch of the library. 

If you want to provide feedback on the new API, ask questions or suggest changes: Now is your chance 😉

Note that the new API is a bit different than the one outlined in the previous API draft issue.

**Comment by jaliph** [comment]
Kindly keep the concept of Thread Pool so in this release. It is very crucial use case, where we can have a constant thread pool running, and assign task in dynamic payload.

**Comment by andywer** [comment]
@jaliph The thread pool is already re-implemented, but the feature branch is not merged yet.

Branch: [`feature/v1-pool`](https://github.com/andywer/threads.js/tree/feature/v1-pool)

Check out *Code Samples > Thread pool* in the feature branch readme 😉
What do you think?

**Comment by jaliph** [comment]
@andywer  This is amazing. I hope we are not missing any of the features provided in earlier version.

Some sample Benchmarks might help in increasing the reliability.

What is the plan for releasing this latest version? Any particular dates you have in mind?

**Comment by andywer** [comment]
In theory v1 should ship all previous features, just dropping the support for inline function workers.

No concrete date yet, but the most complex and difficult things are mostly done and working now, so the rest shouldn't take too much time.

I will merge the pool support in the next few days and it doesn't work with webpack out of the box, but I have already prepared a simple [zero-config webpack plugin](https://github.com/andywer/threads-plugin) that will make it work :)

Still need to ensure that the webpack plugin creates working backend bundles as well as browser bundles.

**Comment by aarmora** [comment]
Hey @andywer! I decided to turn my constant twitter barrage into official stuff here. I'm loving v1 and I have been messing with Pool today. The events logged say that the size is 2 in my case but I don't seem to see any queueing.

I would have expected it to show two blocks of `all the items`, output the numbers looped through for those two arrays, and then two more blocks of `all the items`. Instead it looks like it's running everything immediately.

Example code:

```
// Index.ts

import { spawn, Thread, Worker, Pool } from 'threads';

(async () => {
    let spawnCalled = 0;
    const spawnLooper = () => {
        spawnCalled++;
        return spawn<(items: any[]) => Promise<any>>(new Worker('./../../../../dist/looper.js'));
    };

    const pool = Pool(spawnLooper, 2);

    pool.events().subscribe(console.log);

    const bigList = new Array(100);
    const listWithNumbers: any[] = [];

    for (let i = 0; i < bigList.length; i++) {
        listWithNumbers.push(i);
    }
    console.log('before');

    await pool.queue(async looper => {
        const promises: any[] = [];
        for (let i = 0; i < listWithNumbers.length; i++) {
            promises.push(looper(listWithNumbers.slice(i, i + 25)));
            i += 25;
        }
        await Promise.all(promises);
    });

    await pool.terminate();
    console.log('after', spawnCalled);

})();
```

```
// looper.ts
import { expose } from 'threads/worker';

expose(async (items: any[]) => {
    console.log('all the items', items);

    for (let i = 0; i < items.length; i++) {
        await wait(items[i]);
    }

    function wait(i: number) {
        return new Promise((resolve) => {
            setTimeout(() => {
                console.log(i);
                resolve();
            }, 150);
        })
    }
    return Promise.resolve();
});
```


Image Example:
![image](https://user-images.githubusercontent.com/6865538/59132346-33af3000-8932-11e9-848f-2416dd6c2b4b.png)

**Comment by PolyPik** [comment]
> Provide a convenient way to stream results

Have you considered [readable-stream](https://www.npmjs.com/package/readable-stream)?

**Comment by andywer** [comment]
> Have you considered readable-stream?

Not actively, no. I am preparing a PR introducing an example of how to stream partial results using observables, since they are already supported. But you do have a point with readable streams, nevertheless...

Let's see if more people require readable stream support and implement it then.

**Comment by PolyPik** [comment]
It is just that when you mention streaming, I expected something like the streams in nodejs.

**Comment by deanius** [comment]
I'm excited for a 1.0 - you have a 'finish the documentation' task in this one but I thought I'd also mention that the currently published version has a broken link to the source code in its package.json. Do you want that as a separate issue or just call it a subtask of 1.0?

**Comment by andywer** [comment]
> that the currently published version has a broken link to the source code in its package.json

Ohh, I didn’t notice! Thanks, will fix it straight away 👍

**Comment by andywer** [comment]
**What has been published on Sunday, though, is v1.0 beta!**

Go and give it a try. Should be quite convenient to use. In case you have got any troubles with it, share them here and they shall be fixed when the final v1.0.0 goes live 😉

The new documentation site is now live, too: https://threads.js.org

Repo link in package.json is fixed on master, but not yet published.

**Comment by andywer** [comment]
All the necessities have been covered. Time for the big release…
